pypesto/objective/history.py

```diff
@@ -205,38 +208,137 @@ def start_time(self) -> float:
 
     def get_x_trace(self) -> Sequence[np.ndarray]:
         """Parameter trace."""
+        return [self.get_x(iteration)
+                for iteration in range(len(self))]
+
+    def get_x(self, iteration: int) -> np.ndarray:
+        """Parameter value at iteration
+
+        Parameters
+        ----------
+        iteration:
+            iteration index
+        """
         raise NotImplementedError()
```

Comment:
I think have these 2 functions is not optimal. x_trace in its default implementation is in general quite inefficient and thus must be overwriten in all subclasses. I think (we also discussed internally shortly) that it may be better to give `get_x_trace(self, iterations: Union[Sequence[int], int])` the iteration argument (yes, breaking former API :grimacing:), maybe even allowing for a list of ints.

Reply:
Yes, had that implementation at some point. I don't see the benefit of having this as a single function. If there is a more efficient implementation, the generic one can be overwritten, as currently done for FVAL, CHI2 and TIME in CsvHistory (for everything else I don't see a more efficienty way of implementing this). 

Passing a sequence is fine and its easy to extend `get_{...}_trace` accordingly, but implementing all of that in a single function is just messy in terms of typehints and implementation (you need to add a check whether iterations is a sequence/single int, you need to postprocess the result based on that, etc). I just didn't see a way to write less/simpler code by implementing both functionalities in a single function.

Reply:
As example for X/MemoryHistory
```
if ixs is None:
    return self._trace[X]
else:
   if len(ixs) == 1:
       return self._trace[X][ixs]
   else:
       return [self._trace[X][ix] for ix in ixs]
```

You would have three independen implementation in one function without any reuse and I don't see any improvement to maintainability. With the seperate implementation you could at least reuse the `self._trace[X][ixs]` part, although `[val for ix, val in enumerate(self._trace[X]) if ix in ixs]` is probably more efficient. I don't see how this is fixable by decorators.